File: algorithms/python/3.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        #position of every char
        if len(s) <2:
            return len(s)
        pos = {}
        max_length = 0
        start = 0
        for i in range(len(s)):
            if s[i] not in pos:
                pos[s[i]] = i
            else:
                start = max(pos[s[i]] + 1,start)
                pos[s[i]] = i
            max_length = max(max_length,i-start + 1)
        return max_length

File: algorithms/cpp/3.cpp
#include "iostream"
#include "string"
#include "map"
#include "iterator"
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
	if (s.size()<2)
	{
		return s.size();
	}
	int max_lenght = 0;
	int start = 0;
	map<char,int> dict;
	for(int i =0;i<s.size();i++)
	{
		map<char,int>::iterator iter = dict.find(s[i]);
		if(iter != dict.end())
		{
			start = max(start,dict[s[i]] + 1);
			dict[s[i]] = i;
		}
		else
		{
			dict[s[i]] = i;
		}
		max_lenght = max(max_lenght,i-start+1);
	}
	return max_lenght;
    }
};

int main(int argc, char const *argv[])
{
	string test = "abba";
	Solution *pObj = new Solution();
	cout<<pObj->lengthOfLongestSubstring(test)<<endl;
	delete pObj;
	return 0;
}